internal/otelarrow/admission2/boundedqueue.go

```diff
@@ -0,0 +1,177 @@
+	}
+}
+
+// acquireOrGetWaiter returns with three distinct conditions depending
+// on whether it was accepted, rejected, or asked to wait.
+//
+// - element=nil, error=nil: the fast success path
+// - element=nil, error=non-nil: the fast failure path
+// - element=non-nil, error=non-nil: the slow success path
+func (bq *BoundedQueue) acquireOrGetWaiter(pending uint64) (*list.Element, error) {
+	bq.lock.Lock()
+	defer bq.lock.Unlock()
+
+	if pending > bq.maxLimitAdmit {
+		// when the request will never succeed because it is
+		// individually over the total limit, fail fast.
+		return nil, ErrRequestTooLarge
+	}
+
+	if bq.currentAdmitted+pending <= bq.maxLimitAdmit {
+		// the fast success path.
+		bq.currentAdmitted += pending
+		return nil, nil
+	}
+
+	// since we were unable to admit, check if we can wait.
+	if bq.currentWaiting+pending > bq.maxLimitWait {
+		return nil, ErrTooMuchWaiting
+	}
+
+	// otherwise we need to wait
+	return bq.addWaiterLocked(pending), nil
+}
+
+// Acquire implements Queue.
+func (bq *BoundedQueue) Acquire(ctx context.Context, pending uint64) (ReleaseFunc, error) {
+	element, err := bq.acquireOrGetWaiter(pending)
+	parentSpan := trace.SpanFromContext(ctx)
+	pendingAttr := trace.WithAttributes(attribute.Int64("pending", int64(pending)))
+
+	if err != nil {
+		parentSpan.AddEvent("admission rejected (fast path)", pendingAttr)
+		return noopRelease, err
+	} else if element == nil {
+		parentSpan.AddEvent("admission accepted (fast path)", pendingAttr)
+		return bq.releaseFunc(pending), nil
+	}
+
+	parentSpan.AddEvent("enter admission queue")
+
+	ctx, span := bq.tracer.Start(ctx, "admission_blocked", pendingAttr)
+	defer span.End()
+
+	waiter := element.Value.(*waiter)
+
+	select {
+	case <-waiter.notify.Chan():
+		parentSpan.AddEvent("admission accepted (slow path)", pendingAttr)
+		return bq.releaseFunc(pending), nil
+
+	case <-ctx.Done():
+		bq.lock.Lock()
+		defer bq.lock.Unlock()
+
+		if waiter.notify.HasBeenNotified() {
+			// We were also admitted, which can happen
+			// concurrently with cancellation. Make sure
+			// to release since no one else will do it.
+			bq.releaseLocked(pending)
+		} else {
+			// Remove ourselves from the list of waiters
+			// so that we can't be admitted in the future.
+			bq.removeWaiterLocked(pending, element)
+			bq.admitWaitersLocked()
+		}
+
+		parentSpan.AddEvent("admission rejected (canceled)", pendingAttr)
+		return noopRelease, status.Error(grpccodes.Canceled, context.Cause(ctx).Error())
+	}
+}
+
+func (bq *BoundedQueue) admitWaitersLocked() {
+	for bq.waiters.Len() != 0 {
+		// Ensure there is enough room to admit the next waiter.
+		element := bq.waiters.Back()
+		waiter := element.Value.(*waiter)
+		if bq.currentAdmitted+waiter.pending > bq.maxLimitAdmit {
+			// Returning means continuing to wait for the
+			// most recent arrival to get service by another release.
```

Comment:
Why don't we re-add the element to `bq.waiters`?

Reply:
We haven't removed the item from the waiters list at this point. The call to removeWaiterLocked() does the removal. At this point, the call to `Back()` on line 136 is peeking at the last (most recently inserted) element. 

Returning at this point means allowing the most-recent request to prevent admission to older requests for which there is potentially resource availability. This is an important step whether we use FIFO or LIFO--regardless of the order we assign the requests in the queue, if the first element can't be granted and we move on the subsequent elements, there's a chance of resource starvation--that a large request could stay pending forever. The return here, therefore, ensures priority to the first-out request.

Reply:
Thanks for clarifying I didn't realize `Back` peeked but didn't pop